crypto/fipsmodule/evp/p_hmac.c

```diff
@@ -0,0 +1,230 @@
+ *    prior written permission. For written permission, please contact
+ *    licensing@OpenSSL.org.
+ *
+ * 5. Products derived from this software may not be called "OpenSSL"
+ *    nor may "OpenSSL" appear in their names without prior written
+ *    permission of the OpenSSL Project.
+ *
+ * 6. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by the OpenSSL Project
+ *    for use in the OpenSSL Toolkit (http://www.OpenSSL.org/)"
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
+ * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ *
+ * This product includes cryptographic software written by Eric Young
+ * (eay@cryptsoft.com).  This product includes software written by Tim
+ * Hudson (tjh@cryptsoft.com). */
+
+#include <openssl/evp.h>
+
+#include <openssl/asn1.h>
+#include <openssl/err.h>
+#include <openssl/hmac.h>
+#include <openssl/mem.h>
+#include <openssl/obj.h>
+
+#include "../../evp_extra/internal.h"
+#include "../digest/internal.h"
+#include "internal.h"
+
+
+typedef struct {
+  const EVP_MD *md; /* MD for HMAC use */
+  uint8_t *key;
+  size_t key_len;
+  HMAC_CTX ctx;
+} HMAC_PKEY_CTX;
+
+static int hmac_init(EVP_PKEY_CTX *ctx) {
+  HMAC_PKEY_CTX *hctx;
+  hctx = OPENSSL_malloc(sizeof(HMAC_PKEY_CTX));
+  if (hctx == NULL) {
+    return 0;
+  }
+  OPENSSL_memset(hctx, 0, sizeof(HMAC_PKEY_CTX));
+  HMAC_CTX_init(&hctx->ctx);
+  ctx->data = hctx;
+  return 1;
+}
+
+static int hmac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src) {
+  HMAC_PKEY_CTX *sctx, *dctx;
+  if (!hmac_init(dst)) {
+    return 0;
+  }
+  sctx = src->data;
+  dctx = dst->data;
+  dctx->md = sctx->md;
+  HMAC_CTX_init(&dctx->ctx);
+  if (!HMAC_CTX_copy_ex(&dctx->ctx, &sctx->ctx)) {
+    return 0;
+  }
+
+  if (sctx->key_len != 0) {
+    dctx->key = OPENSSL_memdup(sctx->key, sctx->key_len);
+    if (dctx->key == NULL) {
+      return 0;
+    }
+    dctx->key_len = sctx->key_len;
+  }
+  return 1;
+}
+
+static void hmac_cleanup(EVP_PKEY_CTX *ctx) {
+  HMAC_PKEY_CTX *hctx = ctx->data;
+  if (hctx == NULL) {
+    return;
+  }
+  HMAC_CTX_cleanse(&hctx->ctx);
+  if (hctx->key != NULL) {
+    OPENSSL_free(hctx->key);
+  }
```

Comment:
Our free checks for null already.
```suggestion
OPENSSL_free(hctx->key);
```